nomad-base/src/merkle_sync.rs

```diff
@@ -0,0 +1,151 @@
+    /// DB with home name key prefix
+    pub db: NomadDB,
+}
+
+/// IncrementalMerkleSync errors
+#[derive(Debug, thiserror::Error)]
+pub enum IncrementalMerkleSyncError {
+    /// IncrementalMerkleSync receives ChainCommunicationError from chain API
+    #[error(transparent)]
+    ChainCommunicationError(#[from] ChainCommunicationError),
+    /// DB Error
+    #[error("{0}")]
+    DbError(#[from] DbError),
+}
+
+impl IncrementalMerkleSync {
+    /// Instantiate new IncrementalMerkleSync
+    pub fn default(db: NomadDB) -> Self {
+        Self {
+            merkle: Arc::new(RwLock::new(Default::default())),
+            db,
+        }
+    }
+
+    /// Instantiate new IncrementalMerkleSync from DB
+    pub fn from_disk(db: NomadDB) -> Self {
+        let mut tree = IncrementalMerkle::default();
+        let mut last_committed_root = H256::default();
+
+        if let Some(latest_root) = db.retrieve_latest_committed_root().expect("db error") {
+            for i in 0.. {
+                match db.message_by_leaf_index(i) {
+                    Ok(Some(message)) => {
+                        debug!(
+                            leaf_index = i,
+                            last_committed_root = ?last_committed_root,
+                            "Ingesting leaf from_disk"
+                        );
+
+                        tree.ingest(message.leaf());
+                        last_committed_root = message.committed_root;
+
+                        if tree.root() == latest_root {
+                            break;
+                        }
+                    }
+                    Ok(None) => break,
+                    Err(e) => {
+                        error!(error = %e, "Error in IncrementalMerkleSync::from_disk");
+                        panic!("Error in IncrementalMerkleSync::from_disk");
+                    }
+                }
+            }
+
+            info!(target_latest_root = ?latest_root, last_committed_root = ?last_committed_root, root = ?tree.root(), "Reloaded IncrementalMerkleSync from_disk");
+        }
+
+        Self {
+            merkle: Arc::new(RwLock::new(NomadIncrementalMerkle::new(
+                tree,
+                last_committed_root,
+            ))),
+            db,
+        }
+    }
+
+    /// Start syncing merkle tree with DB
+    pub fn sync(&self) -> Instrumented<JoinHandle<Result<()>>> {
+        let merkle = self.merkle.clone();
+        let db = self.db.clone();
+
+        let span = info_span!("IncrementalMerkleSync");
+        tokio::spawn(async move {
+            loop {
+                let tree_size = merkle.read().await.count();
+
+                info!("Waiting for leaf at index {}...", tree_size);
+                let message = db.wait_for_message(tree_size as u32).await?;
+
+                info!(
+                    index = tree_size,
+                    leaf = ?message.leaf(),
+                    "Ingesting leaf at index {}. Leaf: {}.",
```

Comment:
we're phasing out formatting in these messages. Remove the `{}` in favor of just setting the pre-message params